### ISSUE TITLE
Search or move selection should prefer (shorter) mates

### ISSUE THREAD

**Issue by Mardak** [LeelaChessZero/lc0#837]
Not implemented in #822 but exists in #700, I quickly put together a buggy implementation to explore what's there to consider, and what is actually meaningful to implement.

![Screen Shot 2019-04-22 at 10 26 44 AM](https://user-images.githubusercontent.com/438537/56514485-412b6a80-64ea-11e9-8a40-3ece9e4d6cba.png)
```
position fen 7k/6pp/5p2/8/5R1Q/8/7K/8 b - - 0 1 moves g7g6 h4f6 h8g8

f4e4 N:    9282 (P:  3.87%) (Q:  1.00000) (U: 0.05047) (Q+U:  1.05047) (V:  1.0000) (T101) 
h2h3 N:   15911 (P:  6.63%) (Q:  1.00000) (U: 0.05047) (Q+U:  1.05047) (V:  1.0000) (T101) 
f6e6 N:   44797 (P: 36.96%) (Q:  0.99995) (U: 0.10002) (Q+U:  1.09997) (V:  0.9984) 
f6e7 N:  108277 (P:  1.32%) (Q:  1.00000) (U: 0.00147) (Q+U:  1.00147) (V:  1.0000) (T3) 
f6d8 N:  121259 (P:  1.48%) (Q:  1.00000) (U: 0.00147) (Q+U:  1.00147) (V:  1.0000) (T3) 
f6f7 N:  252327 (P:  2.03%) (Q:  1.00000) (U: 0.00097) (Q+U:  1.00097) (V:  1.0000) (T2) 
f6f8 N:  519160 (P:  2.03%) (Q:  1.00000) (U: 0.00047) (Q+U:  1.00047) (V:  1.0000) (T1) 
```

Here I embedded distance in the node's Q value by multiplying Q with the distance, e.g., q_ == 1.0f for a move is checkmate and q_ == -2.0f is checkmated in 2 (ply), and exposed `GetQ` as bounded [-1, 1] and added a `GetDistance` for use in Backup and PickNode and VerboseStats. The actual implementation will probably be different when optimizing for performance of the desired feature.

The more interesting question is what do we actually want?

If we have search focus on moves that have a mate/terminal win, depending on how much that is biased, it can prevent finding an even shorter mate. E.g., above h2h3 has the highest prior (using 50783) of the terminals and preferring it could have prevented search from realizing the lower prior move f6f8 is checkmate.

On the flip side, allowing search to explore more even with a known mate could result in search still directing too many visits to moves with high priors, e.g.,
```
position fen 7k/6pp/5p2/8/5R1Q/8/7K/8 b - - 0 1 moves g7g6 h4f6 h8g8 f6e6 g8g7

e6g6 N:     132 (P:  1.12%) (Q:  1.00000) (U: 0.06639) (Q+U:  1.06639) (V:  1.0000) (T100) 
e6f7 N:     175 (P:  1.50%) (Q:  1.00000) (U: 0.06749) (Q+U:  1.06749) (V:  1.0000) (T102) 
f4f7 N:     184 (P:  1.62%) (Q:  1.00000) (U: 0.06904) (Q+U:  1.06904) (V:  1.0000) (T105) 
e6f6 N:     541 (P:  1.33%) (Q:  1.00000) (U: 0.01932) (Q+U:  1.01932) (V:  1.0000) (T5) 
e6e7 N:    6863 (P:  1.34%) (Q:  1.00000) (U: 0.00154) (Q+U:  1.00154) (V:  1.0000) (T2) 
e6d6 N:   16134 (P: 38.51%) (Q:  1.00000) (U: 0.01885) (Q+U:  1.01885) (V:  1.0000) (T4) 
```

Alternatively, if picking shorter mates is done only for selecting the best move instead of search, this then wouldn't help improve training data (over what would happen naturally with `--stick-endgames` bumping the eval to +1/0/-1). If this is actually preferred for "cosmetic" play style, then the implementation would probably be quite different focusing on `GetBestChildrenNoTemperature` doing a "one-time" dynamic traversal for shorter mate instead of continuously doing it during search.

I'm sure @Videodr0me has additional thoughts. 😄

**Comment by Mardak** [comment]
To expand a bit more on searching wider vs focusing on a known terminal, I happened to play around with this by setting an artificial Q for checkmate move then linearly decreasing based on the distance without affecting U. For example, if checkmate is Q = 1.1 during search and a TB win is treated as a distance of 100, a linear decrease of 0.0005f results in a #1 = 1.1, #2 = 1.0995, #100/TB win = 1.05, TBwin-in-100 = 1.0f.

These magic numbers bias search by allowing U of other moves to get visits or not get visits if Q difference is big enough. E.g., making checkmate Q = 2.0 would make it much harder for non-terminals to get any additional visits.

As pointed out in the original comment, multiple terminal moves are still heavily biased towards highest prior moves as their U component would generally be much larger, so there might need to be changes to U and/or P to get search to favor shorter mates.

**Comment by Mardak** [comment]
Perhaps the question for @mooskagh and/or @Tilps is do we only want the play style to improve by picking the shorter mate move or do we also want search to actually change visit distribution affecting training data?

Neither should affect Elo (unless we suspect there to be generalizations made by the network to play stronger in favoring moves that look like shorter mates) as search would have found a "terminal" win anyway. I suppose an argument that there is Elo improvement is if we expect lc0 to fail to convert wins because it preferred longer mates and ran into time trouble… ?

**Comment by chanep** [comment]
What about this more general aproach. Instead of linear decrease for mates, a factor (something close to 1, e.g. 0.999) to apply to leaf value in every Backup step. So this works for mates to opponent, mates to us and also for good and bad positions in general.  Search will prefer moves that ends in distant bad positions (or mates to us) instead of moves that ends in a same valued position but with a shorter path. In the same way Search will prefer the opposite for good positions

**Comment by Videodr0me** [comment]
A couple of notes:

I. Encoding the distance to mate allows leela to prefer shorter mates over longer mates, prevents postponing mate-in-ones and allows leela to more accurately display mate (even more so than CP700). The most efficient way is to overload Q with the distance to mate by  the using the Q = 1.x - 0.00xx * distance_to_mate (limited to >=1), because  
- It assures that if q_1 > q_2 we really prefer q1 at all times, which is not the case if you multiply q by distance to mate, then you have to always convert q (or at least check if it needs conversion) in the getter.
- there are many more normal nodes that are not proven mates so always converting q is really inefficient especially as it needs to be done in picknodetoextend to be cpuct compatible
- setting it in a cpuct compatible way on creation of a proven win (or loss) avoids all future conversions and retains consistency and compatibility.
- chaneps proposal would break linear consistency (see below) of the full tree. 

The drawback is that we do have q that are >= 1 and <= 1, which might alieanate some more conversative devs (which is why i didn't add that to CP700). Also the consistency of the tree's qs (as a linear combination of children) is now restricted to non-terminal nodes only, but that is  good enough for other potential enhancements like "turbo-certainty" propagation (which i did not even mention in my comparision of PR822 and PR700).

II. Preferring real terminal wins (no-children) over just certain wins (here terminals with children) in gettingthebestmove over everything else regardless of visits (analogous logic for losses). This was my solution without overloading q to reduce the trolling by postponing mates. It works suprisingly well, as generally shorter mates are proven first and tend to have more visits, but it sometimes fails. Overloading q with the distance to mate is better in that respect, but unfortunately it still makes sense to change bestmove selection strategy even with a distance to mate overloading, because without:
- a proven win might not be played because its visits did not catch up
- a proven loss is not avoided because the visits of an alternative did not overtake it
- a proven draw is not played even if the highest n move as a q<0 if its visits did not catch up.

All these cases are handled by CP700 but the main drawback is, that there are not compatible with training as such. The best move might not have the most visits. But its possible to add some logic to even improve training. If for example we just adjust the visits by those roles, so that the best move(s) get the most visits assigned, it should improve policy learning. It is a complication though, and i would love to see CP be used in training sooner rather than later, so i would maybe leave this for later.

-

**Comment by Videodr0me** [comment]
> As pointed out in the original comment, multiple terminal moves are still heavily biased towards highest prior moves as their U component would generally be much larger, so there might need to be changes to U and/or P to get search to favor shorter mates.

I would not do that (adjusting P or U). If a branch of the tree is a proven win, that should get backpropped as soons as possible, especially if the same proven terminal win is hit repeatedly. It takes no NN eval and is very informative. Adjusting P or U to explore alternatives to that win, will involve at least some NN evals, which might be needed in other parts of the tree. Theoreticalyl searching for searcher mates when there is no need to will most certainly loose elo. Yes, sometimes policy might still prefer a longer mate even with the 1.x - 0.00y*distance_mate_encoding but in the vast majority of cases it will converge and even if it doesn't, i would hate to loose elo just to find shorter mates.

**Comment by gonzalezjo** [comment]
After you find a proven mate--say a mate in 13--you can store the move(s) that guarantees it in some variable. You can then artificially limit search depth to twelve. The next time a mating move is found, overwrite the variable with the new mating move(s) and lower the search depth even more. 


This gets you “allowing search to explore more even with a known mate“ while actually directing nodes where they matter more (towards search for a quicker win)

**Comment by Videodr0me** [comment]
> After you find a proven mate--say a mate in 13--you can store the move(s) that guarantees it in some variable. You can then artificially limit search depth to twelve. The next time a mating move is found, overwrite the variable with the new mating move(s) and lower the search depth even more.

If you find a proven mate in x, in normal play, you should just play it. In sudden death time controls it can be dangerous to continue searching, especially since our latency is much higher than that of a/b engines.

In infinite search one might use such a scheme to find shorter mates if multi-pv is deactivated. With multi-pv I found it much more informative to continue search the other moves, rather than include the already established mating move.

**Comment by Mardak** [comment]
A potentially more general solution is to have the network learn to predict how many moves are left for a position. This is a trainable target given a final move count and moves that led up to it (caveats with adjudication). With the additional moves-left head, search could be biased towards/away from shorter games, e.g., shorter games, smarter time management, contempt.

**Comment by Mardak** [comment]
#992 practically gets us to preferring shorter mates in many cases as shorter mates are usually found earlier in search and could be the only mating move found anyway. If #961 lands with adding a moves `M`/`moves_left` on each node where the NN can estimate that value, similar to sticky-endgames propagating a terminal Q value to replace the NN/search estimated Q, we can backup known values of M.
